src/lib/src/x86_arrops.cpp

```diff
@@ -0,0 +1,133 @@
+}
+
+constexpr std::size_t kBigVectorThreshold = 256 * 1024;
+constexpr std::size_t kStreamingThreshold = 16 * 1024 * 1024;
+
+template<bool streaming_store, typename VectorT, typename FloatingT>
+void CalcBondsOrthoX86Vec(const FloatingT* coords0, const FloatingT* coords1,
+                          const FloatingT* box, std::size_t n,
+                          FloatingT* out) {
+  constexpr std::size_t vector_len = sizeof(VectorT) / sizeof(FloatingT);
+  FloatingT boxx = box[0], boxy = box[1], boxz = box[2];
+  VectorT boxa, boxb, boxc;
+  Interleave3(boxx, boxy, boxz, boxa, boxb, boxc);
+
+
+  size_t i = 0;
+  for (; distopia_unlikely(!IsAligned<VectorT>(&out[i]) && i < n); ++i) {
+    auto x0 = coords0[3 * i], x1 = coords1[3 * i];
+    auto y0 = coords0[3 * i + 1], y1 = coords1[3 * i + 1];
+    auto z0 = coords0[3 * i + 2], z1 = coords1[3 * i + 2];
+    out[i] = Distance3DWithBoundary(x0, y0, z0, x1, y1, z1, boxx, boxy, boxz);
+  }
+  for (; i + vector_len - 1 < n; i += vector_len) {
+    auto j0 = 3 * i;
+    auto j1 = j0 + vector_len;
+    auto j2 = j1 + vector_len;
+
+    auto a0 = loadu_p<VectorT>(&coords0[j0]);
+    auto a1 = loadu_p<VectorT>(&coords1[j0]);
+    auto b0 = loadu_p<VectorT>(&coords0[j1]);
+    auto b1 = loadu_p<VectorT>(&coords1[j1]);
+    auto c0 = loadu_p<VectorT>(&coords0[j2]);
+    auto c1 = loadu_p<VectorT>(&coords1[j2]);
+
+    auto da = Distance1DWithBoundary(a0, a1, boxa);
+    auto db = Distance1DWithBoundary(b0, b1, boxb);
+    auto dc = Distance1DWithBoundary(c0, c1, boxc);
+
+    VectorT dx, dy, dz;
+    Deinterleave3(da, db, dc, dx, dy, dz);
+
+    auto res = Hypot(dx, dy, dz);
+    if constexpr (streaming_store) {
+      stream_p(&out[i], res);
+    } else {
+      store_p(&out[i], res);
+    }
+  }
+  for (; distopia_unlikely(i < n); ++i) {
+    auto x0 = coords0[3 * i], x1 = coords1[3 * i];
+    auto y0 = coords0[3 * i + 1], y1 = coords1[3 * i + 1];
+    auto z0 = coords0[3 * i + 2], z1 = coords1[3 * i + 2];
+    out[i] = Distance3DWithBoundary(x0, y0, z0, x1, y1, z1, boxx, boxy, boxz);
+  }
+  if constexpr (streaming_store) {
+    _mm_mfence();
+  }
+}
+
+template<typename T> struct SmallVecTStruct;
+template<> struct SmallVecTStruct<float> { using type = __m128; };
+template<> struct SmallVecTStruct<double> { using type = __m128d; };
+template<typename T> using SmallVecT = typename SmallVecTStruct<T>::type;
+
+template<typename T> struct BigVecTStruct { using type = SmallVecT<T>; };
+#ifdef DISTOPIA_X86_AVX
+  template<> struct BigVecTStruct<float> { using type = __m256; };
+  template<> struct BigVecTStruct<double> { using type = __m256d; };
+#endif
+template<typename T> using BigVecT = typename BigVecTStruct<T>::type;
+
+template<typename T>
+void CalcBondsOrthoDispatch(const T* coords0, const T* coords1,
+                            const T* box, std::size_t n, T* out) {
+  bool out_aligned = reinterpret_cast<std::uintptr_t>(out) % sizeof(T) == 0;
```

Comment:
Again if we force our own allocator may be able to avoid checks for alignment. Not sure how feasible that is.

Reply:
That doesn’t sound very easy at all 😕 Ideally we’d like permit the user to re-use memory that they’ve already allocated (akin to NumPy’s `out` kwarg). One way to reduce code complexity would be to make more assumptions (e.g. every out buffer is 16-byte aligned) and fall back to the scalar version if these are wrong.

Reply:
I think you are right that enforcing our own memory allocation is possibly wasteful, rather than just taking some kind of view as would be done in the numpy style.